### PR TITLE
feat: allow multiple slots to be defined in a custom component [ALT-416]

### PR DESCRIPTION
**Pull request by primeinteger**
## Purpose

Implements the ability to define multiple slots in a custom component that can be used to drop nested children in different slot drop zones.

To enable multiple slots in a custom component, define the `slots` configuration in the component definition including a unique key (used internally as the `slotId`) and a `displayName` (to show in the canvas component label and layers tab)

```ts
defineComponents([
  {
    component: MyCustomComponent,
    definition: {
      // ... other props ...
      slots: {
        childrenSlot1: {
          displayName: 'Slot 1',
        },
        childrenSlot2: {
          displayName: 'Slot 2',
        },
      },
    },
  }
])
```

In the custom component, handle the props for each slot. Use the key of each slot definition object as the prop variable name:

```ts
const MyCustomComponent: React.FC<Props> = ({ childrenSlot1, childrenSlot2, ...props }) => {
  return (
    <div {...props}>
      <div className='left'>
        {childrenSlot1}
      </div>
      <div className='right'>
        {childrenSlot2}
      </div>
    </div>
  );
};
```

When the custom component is added to an experience, the UI will auto-insert a Container component into each slot with the Display Name set to the slot's configured display name.

Here are the related UI changes to accomplish this:
* https://github.com/contentful/user_interface/pull/21206/files

I added a `<NestedSlots />` component to the Test App to demonstrate this new capability.

Screenshot 2024-05-31 at 3 03 55 PM

### CODE REVIEW

**Comment by loweisz** [comment]
@primeinteger Has a decision been made to make this part of the variables? My understanding is that the variables within the component definition should always be the component variables related to design or content that define the look and feel of the component and is something dynamic for the editor to be changed. 

This configuration seems to be more related to the behaviour of the component and therefore would imo fit better right under definition or under options.

**Comment by primeinteger** [comment]
@loweisz I decided to use the `variables` config for the Dropzone definitions because of how variables are passed into the component as props, it just seemed simple and clean to me.  Thanks for raising this point, maybe I should add a new object in the component definition specifically for the Dropzones configuration so that it's separate from the design/content variables.  I'll work on this today

**Comment by loweisz** [comment]
@primeinteger Thanks for following up, it's in the end just my opinion. Maybe we could talk about this as well with the rest in the next sync? Just so that we are all aligned on the structure as this is a non reversible decision after all 😬

**Review by elylucas — APPROVED** [review]
Nicely done! Left a few small comments but nothing to hang up on.

File: packages/visual-editor/src/components/Dropzone/EditorBlock.tsx [hunk 1]
```diff
@@ -55,7 +58,7 @@ export const EditorBlock: React.FC<EditorBlockProps> = ({
   const isAssemblyBlock = node.type === ASSEMBLY_BLOCK_NODE_TYPE;
   const isAssembly = node.type === ASSEMBLY_NODE_TYPE;
   const isStructureComponent = isContentfulStructureComponent(node.data.blockId);
-  const isEmptyZone = !node.children.length;
+  const isEmptyZone = !node.children.filter((node) => node.data.slotId === slotId).length;
```
[the remark this answers is not in the document]
Comment (primeinteger) [reply packages/visual-editor/src/components/Dropzone/EditorBlock.tsx]: Good idea

File: packages/visual-editor/src/utils/zone.ts [hunk 2]
```diff
@@ -27,3 +27,8 @@ export const getZoneParents = (zoneId: string) => {
 
   return getZonesToRoot(element);
 };
+
+export const parseZoneId = (zoneId: string) => {
```
[the remark this answers is not in the document]
Comment (primeinteger) [reply packages/visual-editor/src/utils/zone.ts:38]: I added comments above the function to provide more detail

**Review by elylucas — APPROVED** [review]

**Pull request merged**